### PR TITLE
Increase reliability when grid engine CLI tools hang or return invalid data

### PR DESCRIPTION
**Pull request by mdpearson**
Hi Erik,

This PR fixes a number of prod issues we've seen with grid engine in the past quarter or so.

The big one is when qstat falsely returns no output. If qstat returns nothing, this PR will sleep 30 sec and then retry.

A second issue is when GE commands hang indefinitely. This PR uses subprocess.run() to enforce timeouts.

There are logging improvements, esp. when Cosmos is waiting for a long time for jobs to complete. And I've made the teardown code more consistent by making a base implementation of `cleanup_task()`. 

I'm still testing this but it's working well enough I thought I'd file a PR and garner your thoughts.

In an attempt to be a good neighbor, I updated the slurm module to use the new commands I wrote for grid engine.

### CODE REVIEW

File: cosmos/job/drm/DRM_Base.py [hunk 1]
```diff
@@ -67,7 +68,8 @@ def kill_tasks(self, tasks):
             self.kill(t)
 
     def cleanup_task(self, task):
-        pass
+        if task.drm_jobID and task.status not in completed_task_statuses:
+            self.kill(task)
```
Comment (mdpearson) [inline cosmos/job/drm/DRM_Base.py]: I hoisted this from `drm_k8s_jobs.py`, whose implementation seemed more broadly applicable

File: cosmos/job/drm/drm_ge.py [hunk 2]
```diff
@@ -301,17 +301,88 @@ def qacct(job_id, num_retries=10, quantum=30, logger=None, log_prefix=""):
...
+            successful_qdels += 1
+        else:
+            undead_job_ids.append(job_id)
+
+    if undead_job_ids:
+        logger.warning(
+            "qdel returned exit code %s, calling on one job_id at a time", returncode
+        )
+
+        for job_id in undead_job_ids:
```
Comment (mdpearson) [inline cosmos/job/drm/drm_ge.py]: what are the pros/cons of using a ThreadPoolExecutor here? pros -- would return more quickly. cons -- would hammer grid engine at a time when we're not sure it's working 100%

Comment (mdpearson) [reply cosmos/job/drm/drm_ge.py]: maybe add a comment to that effect

File: cosmos/job/drm/drm_ge.py [hunk 3]
```diff
@@ -301,17 +301,88 @@ def qacct(job_id, num_retries=10, quantum=30, logger=None, log_prefix=""):
...
+        len(job_ids),
+    )
+    return successful_qdels
+
+
+def qstat(logger=None):
     """
     Return a mapping of job ids to a dict of GE information about each job.
 
+    If qstat hangs or returns an error, wait 30 sec and call it again. Do this
```
Comment (mdpearson) [inline cosmos/job/drm/drm_ge.py]: this comment is false -- if hangs for more than 30 sec, wait 15 sec and retry. update run_cli_cmd invocation with all params to make this self-documenting & fix the comment

File: cosmos/job/drm/drm_ge.py [hunk 4]
```diff
@@ -301,17 +301,88 @@ def qacct(job_id, num_retries=10, quantum=30, logger=None, log_prefix=""):
...
     """
-    try:
-        lines = subprocess.check_output(['qstat'], preexec_fn=exit_process_group).decode().strip().split('\n')
-    except (subprocess.CalledProcessError, OSError):
+    if logger is None:
+        logger = _get_null_logger()
+
+    stdout, _, returncode = run_cli_cmd(["qstat"], interval=30, logger=logger, attempts=3)
+    if returncode != 0:
+        logger.info("qstat returned %s: If GE is offline, all jobs are dead or done")
```
Comment (mdpearson) [inline cosmos/job/drm/drm_ge.py]: this should be a warning

File: cosmos/job/drm/drm_ge.py [hunk 5]
```diff
@@ -342,22 +413,36 @@ def qsub(cmd_fn, stdout_fn, stderr_fn, addl_args=None, drm_name="GE", logger=Non
...
-        logger.error('%s submission to %s (%s) failed with error %s: %s' %
-                     (log_prefix, drm_name, qsub, cpe.returncode, cpe.output.decode().strip()))
-        status = TaskStatus.failed
-    except ValueError:
-        logger.error('%s submission to %s returned unexpected text: %s' %
-                     (log_prefix, drm_name, out))
+
+    stdout, stderr, returncode = run_cli_cmd(
+        '{qsub_cli} "{cmd_fn}"'.format(cmd_fn=cmd_fn, qsub_cli=qsub_cli),
+        attempts=2,
```
Comment (mdpearson) [inline cosmos/job/drm/drm_ge.py]: set this to 1. we don't want to double-submit jobs

File: cosmos/models/Workflow.py [hunk 6]
```diff
@@ -541,10 +545,20 @@ def _run(workflow, session, task_queue):
             else:
                 raise AssertionError('Unexpected finished task status %s for %s' % (task.status, task))
             available_cores = True
+            last_log_timestamp = time.time()
 
         # only commit Task changes after processing a batch of finished ones
         session.commit()
 
+        if last_log_timestamp + 300 < time.time():
```
Comment (mdpearson) [inline cosmos/models/Workflow.py]: this 300 should be a named constant

File: cosmos/job/drm/util.py [hunk 7]
```diff
@@ -105,3 +61,69 @@ def exit_process_group():
      usually it exists, but performs the operation of setsid()."
     """
     return os.setsid()
+
+
+def run_cli_cmd(
+    args,
+    attempts=1,
+    interval=15,
```
Comment (mdpearson) [inline cosmos/job/drm/util.py]: rename this parameter and set it to 30 sec

File: cosmos/models/Workflow.py [hunk 8]
```diff
@@ -416,10 +416,13 @@ def terminate(self, due_to_failure=True):
 
     def cleanup(self):
         if self.jobmanager:
-            self.log.info('Cleaning up {num_dead_tasks} dead tasks'.format(
-                num_dead_tasks=len(self.jobmanager.dead_tasks),
-            ))
+            self.log.info(
+                "%s Cleaning up %d dead tasks",
```
Comment (mdpearson) [inline cosmos/models/Workflow.py]: "dead tasks" is misleading in this log message. just say "tasks"

File: cosmos/job/drm/drm_ge.py [hunk 9]
```diff
@@ -301,17 +301,88 @@ def qacct(job_id, num_retries=10, quantum=30, logger=None, log_prefix=""):
...
+        logger = _get_null_logger()
+
+    stdout, _, returncode = run_cli_cmd(["qstat"], interval=30, logger=logger, attempts=3)
+    if returncode != 0:
+        logger.info("qstat returned %s: If GE is offline, all jobs are dead or done")
+        return {}
+    lines = stdout.strip().split("\n")
+    if not lines:
+        logger.info(
+            "qstat returned 0, but no output: all jobs are probably done, although GE may be offline"
```
Comment (mdpearson) [inline cosmos/job/drm/drm_ge.py]: reword this log info for clarity

File: cosmos/job/drm/drm_ge.py [hunk 10]
```diff
@@ -301,17 +301,88 @@ def qacct(job_id, num_retries=10, quantum=30, logger=None, log_prefix=""):
...
+    an exit handler that does not have arbitrary amounts of time in which to run.
+    """
+    stdout, stderr, returncode = run_cli_cmd(
+        ["qdel", ",".join(job_ids)],
+        logger=logger,
+        attempts=1,
+        timeout=1,
+    )
+    if returncode == 0:
+        logger.info("qdel reported successful signalling of %d job_ids", len(job_ids))
```
Comment (mdpearson) [inline cosmos/job/drm/drm_ge.py]: "success against"

**Comment by mdpearson** [comment]
@egafni we've tested this pretty thoroughly over here and it should be good to go.

**Comment by egafni** [comment]
Thanks!  Just saw this - could you please resolve the conflicts?

**Comment by egafni** [comment]
Sorry for the delay on merging - please ping me if it takes this long again.  I must have missed the github notification

**Pull request merged**